### ISSUE TITLE
Create a WebExtension

### ISSUE THREAD

**Issue by ExE-Boss** [mozilla/shumway#2435]
# This is being worked on in: [ExE‑Boss/mozilla‑shumway#web‑extension](https://github.com/ExE-Boss/mozilla-shumway/tree/web-extension/extension/webext)

### See&nbsp;also [bug&nbsp;1451083](https://bugzilla.mozilla.org/show_bug.cgi?id=1451083 "[tracking] Support Shumway as a WebExtension")

With XUL, XPCOM and SDK addons being removed in Firefox 57, it would be nice to have a working WebExtension before then.

From my own experimentation, most of the code in `content/web/` is re-useable, except for lines which refer to a URL of a Shumway resource.

Edit: after some testing, this has been resolved and the `browser` object is now accessible, but we should still use relative URLs instead.

Edit 2: this seems to be broken again (<i title="sarcasm"><q>yay…</q></i>), also see [bug 1443253](https://bugzil.la/1443253 "1443253 - WebExtensions: moz-extension files when loaded in iframes get content script scope/privileges").

~~Those get changed to use the following snippet: (which has to be used, because the `browser` namespace is inaccessible from in-page iframes, unless we inject scripts into those through `manifest.json`)~~
```javascript
/**
 * The non-browser specific extension address.
 * @type {string}
 * @author ExE Boss
 */
const extension_url = window.location.protocol + "//" + window.location.host + '/';

/**
 * Converts a relative path within an add-on's install directory to a fully-qualified URL.
 *
 * @param {string} path The path, relative to manifest.json.
 * @return {string} The same result as calling <a href="https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/extension/getURL">
 * <code>browser.extension.getURL(path)</code></a>.
 * @see <a href="https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/extension/getURL">
 * <code>browser.extension.getURL(path)</code></a>
 * @author ExE Boss
 */
function getURL(path) {
  if (path.charAt(0) === '/')
    path = path.slice(1);

  return extension_url + path;
}
```

I have done some work on this at [ExE-Boss/mozilla-shumway#web-extension](https://github.com/ExE-Boss/mozilla-shumway/tree/web-extension), previous attempt has been archived under [ExE-Boss/mozilla-shumway#broken/web-extension/take1](https://github.com/ExE-Boss/mozilla-shumway/tree/broken/web-extension/take1).

**Comment by nanonyme** [comment]
@ExE-Boss Did you ever end up continuing on this? Something like this is definitely necessary if this project is not to be completely forgotten. Also imo if this done, should be investigated what's the amount of effort needed for cross-browser support for Shumway.

**Comment by ExE-Boss** [comment]
I won't be able to dedicate my time to this until at least July, and keep in mind that I also work on [NOVA](https://novaapi.net) and my [own projects](https://ExE-Boss.tech/web-extensions) (which includes a lot of unannounced stuff).

**Comment by ROBERT-MCDOWELL** [comment]
I shared the shumway link to some developers, maybe something good will go out soon.

**Comment by ExE-Boss** [comment]
Also, after some testing, the `browser` object being nonexistent in `moz-extension://` URLs, that aren't the background, options or popup pages, have been resolved in a recent release.

**Comment by ExE-Boss** [comment]
I have shared this on [/r/firefox](https://www.reddit.com/r/firefox/comments/6gnecx/calling_webextension_developers_help_kill_flash/), hopefully some developers will show up and help this project succeed.

**Comment by SoniEx2** [comment]
I see an @ExE-Boss working on a Shumway...

Have you tried talking to Flash devs? e.g. https://twitter.com/MateuszSkutnik creator of http://www.mateuszskutnik.com/submachine/ ?

**Comment by ExE-Boss** [comment]
@SoniEx2 Yes, I have played and won all the Submachine games. (But this discussion about what games I have played is off-topic)

**Comment by SoniEx2** [comment]
I mean go get mateusz to work on shumway

**Comment by ExE-Boss** [comment]
Mateusz is a Flash game developer, not a JavaScript VM developer.

Shumway is a Flash VM implemented entirely in JavaScript using HTML5 technology.

**Comment by romulasry** [comment]
Tired on the latest Firefox 62.0.3, and ran into this issue. It would be interesting if there was an update.

**Comment by mirh** [comment]
> Depends on bug 1457500

Content handlers finally landed in 66 nightly

**Comment by TriMoon** [comment]
Too bad we can't **easily** test the W.I.P. of this extension yet...
Would be nice to have a xpi in the releases section...
